// trace-error/src/Exception.js
import WeakMap from 'weakmap';

const privates = new WeakMap();

export default class Exception {
  static searchPrototype = false;

  get stack() {
    return this._getProperty('error').stack;
  }

  set stack(stack) {
    this._getProperty('error').stack = stack;
  }

  get name() {
    return this._getProperty('error').name;
  }

  set name(name) {
    this._getProperty('error').name = name;
  }

  get message() {
    return this._getProperty('error').message;
  }

  set message(message) {
    this._getProperty('error').message = message;
  }

  // TEMP for extending the base class via Exception.prototype - unsafe procedure
  // marked for deprecation
  get(key) {
    return this._getProperty('error')[key];
  }

  // marked for deprecation
  set(key, value) {
    this._getProperty('error')[key] = value;
  }

  _defineProperty(key, value) {
    const pr = privates.has(this) ? privates.get(this) : {};
    pr[key] = value;
    privates.set(this, pr);
  }

  _getProperty(key, def) {
    return privates.has(this) ? privates.get(this)[key] : def;
  }

  constructor(...args) {
    const error = new Error(...args);
    error.name = this.constructor.name;
    this._defineProperty('error', error);
  }

  toJSON() {
    if (!Exception.searchPrototype) {
      return {stack: this.stack, message: this.message, name: this.name};
    }

    const json = {};
    let currProto = this;
    const chain = [currProto];

    while (currProto = Object.getPrototypeOf(currProto)) {
      chain.push(currProto);
    }

    for (const proto of chain) {
      for (const key of Object.getOwnPropertyNames(proto)) {
        if (json.hasOwnProperty(key)) continue;
        const desc = Object.getOwnPropertyDescriptor(proto, key);
        if (desc) {
          const value = desc.get ? desc.get.call(this) : desc.value;

          if (proto.isPrototypeOf(value)) {
            continue;
          }

          if (typeof value !== 'function') {
            json[key] = value;
          }
        }
      }
    }

    return json;
  }
}

Object.setPrototypeOf(Exception.prototype, Error.prototype);
