// js-to-ts-converter/dist/create-ts-morph-project.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createTsMorphProject = void 0;
const ts_morph_1 = require("ts-morph");
const fast_glob_1 = __importDefault(require("fast-glob"));
/**
 * Creates a ts-morph Project by including the source files under the given
 * `directory`.
 *
 * @param directory The absolute path to the directory of .js files to
 *   include.
 * @param options
 * @param options.indentationText The text used to indent new class property
 *   declarations.
 * @param options.excludePatterns Glob patterns to exclude files.
 */
function createTsMorphProject(directory, options = {}) {
    const tsMorphProject = new ts_morph_1.Project({
        manipulationSettings: {
            indentationText: options.indentationText || ts_morph_1.IndentationText.Tab
        }
    });
    // Read files using fast-glob. fast-glob does a much better job over node-glob
    // at ignoring directories like node_modules without reading all of the files 
    // in them first
    let files = fast_glob_1.default.sync(options.includePatterns || `**/*.+(js|ts|jsx|tsx)`, {
        cwd: directory,
        absolute: true,
        followSymbolicLinks: true,
        // filter out any path which includes node_modules. We don't want to
        // attempt to parse those as they may be ES5, and we also don't accidentally
        // want to write out into the node_modules folder
        ignore: ['**/node_modules/**'].concat(options.excludePatterns || [])
    });
    files.forEach((filePath) => {
        tsMorphProject.addSourceFileAtPath(filePath);
    });
    return tsMorphProject;
}
exports.createTsMorphProject = createTsMorphProject;
