// js-to-ts-converter/dist/logger/logger.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const winston = __importStar(require("winston"));
const winstonLogger = winston.createLogger({
    level: 'verbose',
    transports: [
        new winston.transports.Console({
            format: winston.format.combine(winston.format.colorize(), winston.format.align(), winston.format.printf(info => `${info.level}: ${info.message}`))
        })
    ]
});
/**
 * Abstraction layer for the Winston logger. The methods are in order from
 * highest level of logging to lowest.
 */
class Logger {
    setLogLevel(logLevel) {
        winstonLogger.level = logLevel;
    }
    debug(message) {
        winstonLogger.log('debug', message);
    }
    verbose(message) {
        winstonLogger.log('verbose', message);
    }
    info(message) {
        winstonLogger.log('info', message);
    }
    log(message) {
        winstonLogger.log('info', message);
    }
    warn(message) {
        winstonLogger.log('warn', message);
    }
    error(message) {
        winstonLogger.log('error', message);
    }
}
const logger = new Logger();
exports.default = logger;
