// js-to-ts-converter/dist/converter/correct-js-properties.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const js_class_1 = require("../model/js-class");
const graphlib_1 = require("graphlib");
/**
 * After the graph of original {@link JsClass}es and their properties have been
 * created, we need to remove properties from subclasses that are defined in
 * their superclasses.
 *
 * This function takes the original graph of classes with all properties in each
 * class and returns a new list of JsClasses with the properties properly
 * filtered so that subclasses do not define the same properties that are
 * already present in their superclasses.
 *
 * ## Algorithm
 *
 * 1. Build graph of subclasses -> superclasses
 * 2. Take topological sort of graph
 * 3. Starting at the superclasses in the sort, fill in the
 *    propertySets for each JsClass. For every subclass encountered,
 *    filter out its superclass properties to create the subclass's property
 *    set
 * 4. Use the propertySets to create a new list of JsClasses
 */
function correctJsProperties(jsClasses) {
    const jsClassHierarchyGraph = new graphlib_1.Graph();
    // First, add all nodes to the graph
    jsClasses.forEach(jsClass => {
        jsClassHierarchyGraph.setNode(jsClass.id, jsClass);
    });
    // Second, connect the subclasses to superclasses in the graph
    jsClasses.forEach(jsClass => {
        if (jsClass.superclassId) {
            // As a bit of error checking, make sure that we're not going to
            // accidentally create a graph node by adding an edge to
            // jsClass.superclassId. This would happen if we didn't figure out
            // the correct path to the superclass in the parse phase, or we
            // didn't have the superclass's source file added to the project.
            if (!jsClassHierarchyGraph.hasNode(jsClass.superclassId)) {
                throw new Error(`
					correct-js-properties.ts: no JsClass exists for the 
					superclass identifier: '${jsClass.superclassId}'.
					Was processing JsClass '${jsClass.name}' from path '${jsClass.path}'.
					Make sure that the superclass's .js file is within the 
					directory passed to this conversion utility, or otherwise 
					there was a bug in creating the path/identifier to the superclass. 
				`.replace(/^\t*/gm, ''));
            }
            jsClassHierarchyGraph.setEdge(jsClass.id, jsClass.superclassId);
        }
    });
    // the topological sort is going to put superclasses later in the returned
    // array, so reverse it
    const superclassToSubclassOrder = graphlib_1.alg.topsort(jsClassHierarchyGraph).reverse();
    // Starting from superclass JsClass instances and walking down to subclass
    // JsClass instances, fill in the property sets. When a subclass is
    // encountered, take all of the properties that were used in that subclass,
    // minus the properties in its superclass, in order to determine the
    // subclass-specific properties
    superclassToSubclassOrder.forEach(jsClassId => {
        const jsClass = jsClassHierarchyGraph.node(jsClassId);
        const subclassOnlyProperties = new Set(jsClass.properties);
        const superclasses = getSuperclasses(jsClass);
        superclasses.forEach((superclass) => {
            // Filter out both properties and methods from each superclass
            superclass.members.forEach((superclassProp) => {
                subclassOnlyProperties.delete(superclassProp);
            });
        });
        const newJsClass = new js_class_1.JsClass({
            path: jsClass.path,
            name: jsClass.name,
            superclassName: jsClass.superclassName,
            superclassPath: jsClass.superclassPath,
            methods: jsClass.methods,
            properties: subclassOnlyProperties
        });
        // Re-assign the new JsClass with the correct subclass properties back
        // to the graph for the next iteration, in case there is a subclass of
        // the current class which needs to read those properties
        jsClassHierarchyGraph.setNode(jsClassId, newJsClass);
    });
    // Return all of the new JsClass instances with properties corrected for
    // superclass/subclasses
    return jsClassHierarchyGraph.nodes()
        .map(jsClassId => jsClassHierarchyGraph.node(jsClassId));
    function getSuperclasses(jsClass) {
        const superclasses = [];
        while (jsClass.superclassId) {
            const superclass = jsClassHierarchyGraph.node(jsClass.superclassId);
            superclasses.push(superclass);
            jsClass = superclass;
        }
        return superclasses;
    }
}
exports.correctJsProperties = correctJsProperties;
